src/systems/audio-debug-system.js

```diff
@@ -0,0 +1,323 @@
+  precision highp float;
+  precision highp int;
+  
+  varying vec2 vUv;
+  varying vec3 vNormal;
+  void main()
+  {
+    vNormal = normalize(normal);
+    gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+    vUv = position.xz;
+  }
+`;
+const CONE_FRAG = `
+  // Based on: https://www.shadertoy.com/view/Mll3D4#
+  precision highp float;
+  precision highp int;
+
+  varying vec2 vUv;
+  varying vec3 vNormal;
+  uniform float time;
+
+  const int LINEAR = 0;
+  const int INVERSE = 1;
+  const int EXPONENTIAL = 2;
+          
+  uniform vec3 colorInner;
+  uniform vec3 colorOuter;
+  uniform vec3 colorGain;
+  uniform int count;
+  uniform vec3 sourcePosition[${MAX_DEBUG_SOURCES}];
+  uniform vec3 sourceOrientation[${MAX_DEBUG_SOURCES}];
+  uniform float maxDistance[${MAX_DEBUG_SOURCES}];
+  uniform float refDistance[${MAX_DEBUG_SOURCES}];
+  uniform float rolloffFactor[${MAX_DEBUG_SOURCES}];
+  uniform int distanceModel[${MAX_DEBUG_SOURCES}];
+  uniform float coneInnerAngle[${MAX_DEBUG_SOURCES}];
+  uniform float coneOuterAngle[${MAX_DEBUG_SOURCES}];
+  uniform float gain[${MAX_DEBUG_SOURCES}];
+  uniform bool clipped[${MAX_DEBUG_SOURCES}];
+
+  const float kPi = 3.141592;
+  const float kDegToRad = kPi / 180.0;
+  const float kInvPi = 1.0 / 3.141592;
+  const float kRefDistWidth = 0.05;
+
+  float att_linear(float x, float rolloff, float dref, float dmax) {
+    return 1.0-(rolloff*((x-dref)/(dmax-dref)));
+  }
+
+  float att_inverse(float x, float rolloff, float dref) {
+    return dref/(dref+(rolloff*(max(x, dref)-dref)));
+  }
+
+  float att_exponential(float x, float rolloff, float dref) {
+    return pow((max(x, dref)/dref),-rolloff);
+  }
+
+  vec4 circle(vec2 center, float d, float len, float radius, float holeRadius, vec3 color, float offset) {  
+    // Define how blurry the circle should be. 
+    // A value of 1.0 means 'sharp', larger values
+    // will increase the bluriness.
+    float bluriness = 1.0;
+    
+    // Calculate angle, so we can draw segments, too.
+    float angle = atan( center.x, center.y ) * kInvPi * 0.5;
+    angle = fract( angle + offset);
+    
+    // Create an anti-aliased circle.
+    float wd = bluriness * fwidth( d );
+    float circle = smoothstep( radius + wd, radius - wd, d );
+    
+    // Optionally, you could create a hole in it:
+    float inner = holeRadius;
+    circle -= smoothstep( inner + wd, inner - wd, d );
+    
+    // Or only draw a portion (segment) of the circle.
+    float wa = bluriness * fwidth( angle );
+    float segment = smoothstep( len + wa, len - wa, angle );
+	  segment *= smoothstep( 0.0, 2.0 * wa, angle );
+    circle *= mix( segment, 1.0, step( 1.0, len ) );
+        
+    // Let's define the circle's color now.
+    vec3 rgb = color * circle;
+    
+    // Output final color.
+    return vec4( rgb, circle);
+  }
+
+  vec4 att(float d, vec4 circle, float maxDistance, float refDistance, float rolloffFactor, int distanceModel) {
```

Comment:
Fairly minor since this a debug shader and "computers are fast", but we re-caculate attenuation a few times. 

Should maybe be split into:
```glsl
float att(float d,float maxDistance, float refDistance, float rolloffFactor, int distanceModel)
```
and
```glsl
vec4 draw_att(float attenuation, vec4 circle)
```

Reply:
Good call, updated